### ISSUE TITLE
Feature Request:  sources attribute for proxy clients

### ISSUE THREAD

**Issue by Craigeous** [sensu/sensu#1259]
I've been working with the proxy client and the source attribute in checks.  I am happy they exist, but I'm finding I have to create a check for each proxy client, which is fine, I just wish they provided a little more function.  

I'd like to be able to aggregate the checks as well, which you cannot do(from my knowledge) in the current form.  I was thinking, it would be nice if we had a sources attribute, that had an array of proxy clients, similar to handlers?  Below is an example I was thinking of to illustrate my idea.  Not sure if this is possible or not.

**Register clients:**
`curl -s -i -X POST -H 'Content-Type: application/json' -d '{"name":"switch-x","address":"0.0.0.0","environment": "infrastructure", "subscriptions":[],"keepalives":false,"}' http://sensu-server:4567/clients`

_same for switch-y, firewall-x, firewall-y, router-x, and router-y, etc..._

**Create the checks:**

```
{ 
  "checks": { 
    "check_ping_switch": { 
      "handlers": ["default", "mailer"],
      "command": "check-ping.rb -h :::source.address:::",
      "sources": ["switch-x", "switch-y"],
      "subscribers": ["infrastructure"],
      "interval": 30,
      "aggregate": true 
    },
    "check_ping_firewall": { 
      "handlers": ["default", "mailer"],
      "command": "check-ping.rb -h :::source.address:::",
      "sources": ["firewall-x", "firewall-y"],
      "subscribers": ["infrastructure"],
      "interval": 30,
      "aggregate": true 
    },
    "check_ping_router": { 
      "handlers": ["default", "mailer"],
      "command": "check-ping.rb -h :::source.address:::",
      "sources": ["router-x", "router-y"],
      "subscribers": ["infrastructure"],
      "interval": 30,
      "aggregate": true 
    } 
  } 
} 
```

When the check runs it will look at each entry in source and the address attribute of that client and insert that into the host.  You would have one check/aggregate for each type rather than one for each proxy client, but you'd still have individual clients.

**Comment by lp0** [comment]
I would need :::source.name::: as well as as support for multiple source address families, e.g. a check for SSL using the IPv4 address of the source and a check for SSL using the IPv6 address of the source. Both of these would also require the name of the source to be able to verify the hostname on the certificate.

**Comment by Samgarr** [comment]
+1

**Comment by cwjohnston** [comment]
@Craigeous @lp0  thanks for opening this issue and describing your use cases. I'm tagging this for 
consideration as a future feature. We are definitely interested in reducing the overhead of configuration for checks like you've described here, but the implementation details are still under discussion.

As an aside, starting in [Sensu 0.24](https://sensuapp.org/docs/0.24/overview/changelog.html#core-v0-24-0-changes) we added an `origin` attribute to Sensu event [`check` attributes](https://sensuapp.org/docs/0.25/reference/events.html#check-attributes) which provides the name of the client that generated the check event.

**Comment by absolutejam** [comment]
Funny, I just posted the something similar on Sensu-users (https://groups.google.com/forum/m/#!topic/sensu-users/lGczEs7YC4k). 

This would be really handy as at the minute, each one of my devices requires a single, unique check. This can quickly become tiresome and cause clutter, and defies DRY.

Thanks for taking this into consideration!

**Comment by exeral** [comment]
Here is my usecase that seems related ot this issue

Let's say I have 100 webservers.
I want to check each http://webserverX/status answer HTTP 200 OK
I don't want the webserver subscribe to the http check and check itself because the traffic would remain local
I want an external poller to check the webserver so the whole network chain is validated.

I don't want to setup 100 check, each one a different webserver as source because it would display 100 checks that does the same thing in Uchiwa checks view.

**Comment by cwjohnston** [comment]
In #1555 we're proposing to implement "proxy check requests", a new feature allowing a pubsub check definition to generate a unique check request for each client matching a defined set of client attributes.

Please have a look and share your thoughts?

cc: @Craigeous @lp0 @Samgarr @exeral

**Comment by portertech** [comment]
The proxy check requests feature is now available in the 0.28 release(s) :+1: 

Closing this issue due to inactivity after @cwjohnston's update.
